### ISSUE TITLE
Emulated SaveKey/AtariVox support for Stella libretro core

### ISSUE THREAD

**Issue by dionoid** [stella-emu/stella#1007]
The emulated SaveKey/AtariVox support in the Stella libretro core isn't working correctly.
While (emulated) SaveKey is supported and detected, the data written to the SaveKey is never written back to the `savekey_eeprom.dat` file, so highscores aren't saved across sessions in RetroArch. Note that SaveKey data is currently only written back to non-volatile storage in the destructor of the MT24LC256 class. My guess is that this destructor is never called by the libretro core.

I think saving SaveKey data back to file should be done when the libretro API  `retro_unload_game` is called. I see libretro.cxx has implemented libretro's `retro_unload_game` and `retro_deinit` to do a `stella.destroy()`, which I think basically comes down to a `OSystem.reset();`, but appearently that doesn't destruct the SaveKey controller instance.

**Comment by thrust26** [comment]
`theOSystem.reset()` also causes the execution of the destructor of `MT24LC256`.

![image](https://github.com/stella-emu/stella/assets/24360774/ea30eeb2-cd04-45bf-81ad-e2d975ec76d6)

**Comment by dionoid** [comment]
@thrust26 Ah, thanks for that. My RetroArch log console shows this error:

![image](https://github.com/stella-emu/stella/assets/3540187/65b0197c-e25b-4279-b2f1-455d4db189b6)

**Comment by sa666666** [comment]
There is additional work needed behind the scenes here for Stella to load/save files.  Basically, we don't have a proper implementation of `FSNodeLIBRETRO`, so until that's fixed, none of this can proceed.

**Comment by thrust26** [comment]
Ah, that's what I was already assuming. It seems that Stella is not allowed to write the file. There is not much we can do here.

**Comment by sa666666** [comment]
> Ah, that's what I was already assuming. It seems that Stella is not allowed to write the file.

Not so much 'not allowed', but 'doesn't contain the code to even do it'.

**Comment by thrust26** [comment]
> EDIT: I will add that the libretro port of Stella is very basic, and doesn't expose much configurability nor work with any external files.

And that's why it is a pretty bad idea to base the 2600+ on it.

**Comment by dionoid** [comment]
> And that's why it is a pretty bad idea to base the 2600+ on it.

Libretro frontends (RetroArch, etc.) and OS-based distributions (RetroPi, Batocera) are pretty popular these days, so it hurts to see that the Stella libretro core isn't as functional and configurable as the stand-alone version of Stella.

**Comment by thrust26** [comment]
Maybe RetroArch is popular, but I really do not like its cumbersome UI. 

And like @sa666666 already said, it only supports the basic functions of Stella. Also Libretro is split into two branches, one that is based on pre-4.0 Stella (no hardware acceleration) and a current one. Personally I am not very motivated to do any work for it.

**Comment by thrust26** [comment]
> > Ah, that's what I was already assuming. It seems that Stella is not allowed to write the file.
> 
> Not so much 'not allowed', but 'doesn't contain the code to even do it'.

Oops. That's even worse.

**Comment by sa666666** [comment]
> > And that's why it is a pretty bad idea to base the 2600+ on it.
> 
> Libretro frontends (RetroArch, etc.) and OS-based distributions (RetroPi, Batocera) are pretty popular these days, so it's a pity that the Stella libretro core isn't as functional and configurable as the stand-alone version of Stella.

The extra help we'd hoped to come for the libretro port didn't materialize.  The first pass of porting Stella is done; the second pass was supposed to cover this extra stuff, but it didn't happen.  I am more than willing to guide someone through the process, but I don't personally have the time to do it myself.  At least not in the near future.

**Comment by dionoid** [comment]
> > Libretro frontends (RetroArch, etc.) and OS-based distributions (RetroPi, Batocera) are pretty popular these days, so it's a pity that the Stella libretro core isn't as functional and configurable as the stand-alone version of Stella.
> 
> The extra help we'd hoped to come for the libretro port didn't materialize. The first pass of porting Stella is done; the second pass was supposed to cover this extra stuff, but it didn't happen. I am more than willing to guide someone through the process, but I don't personally have the time to do it myself. At least not in the near future.

At the moment I'm working on porting the [Philips P2000T (M2000) emulator](https://github.com/p2000t/M2000) to Libretro and getting it included in their CI/CD process. As soon as that's done, I'm happy to help out with trying to get the Libretro virtual filesystem to work in Stella. Note that I'm a C/C#/Java/Python software developer, but I never worked with C++ before, so I may need some additional time to get into that. @sa666666, what's the best way to contact you to give me a head start and guide me through the process? 

While I'm also not a huge fan of RetroArch's menu UI, I do think that it's a good idea to get the Stella libretro core to support things like Savekey and other filesystem-related features. Would be nice if these changes end up in the Atari 2600+, but if they don't, that's fine too.

**Comment by thrust26** [comment]
@dionoid Still interested?

**Comment by dionoid** [comment]
> @dionoid Still interested?

Yes, I'm definitely interested in picking this up - with some guidance from @sa666666, of course. I've recently finished porting the P2000T emulator to Libretro (see [M2000 documentation](https://docs.libretro.com/library/m2000/)), so I’ve got some bandwidth now.

@sa666666 Let’s connect after the holidays, if that works for you!

**Comment by sa666666** [comment]
This seems to be fixed when I implemented a proper FSNodeLIBRETRO backend in https://github.com/stella-emu/stella/commit/51994c0fe04fcd28f0220a7462b1f2aaeb0fe12d.  On my system, using the Grizzards ROM, it properly created a directory in `$HOME/.config/retroarch/saves/nvram`, and populated it with `atarivox_eeprom.dat` and `Grizzards_eeprom.dat`.  Upon exiting and reloading the ROM, the AtariVox data was confirmed to be saved and re-loaded.

One thing to mention, though.  The ROM must be known to Stella as using the SaveKey or AtariVox.  If it's built into the properties database, then it works as-is.  Otherwise you need to create a per-ROM properties file.  This has to be next to the ROM, with the same basename, with the extension .pro.  So if the ROM is named `ROM.bin` or `ROM.a26`, the properties file must be named `ROM.pro`.  This is currently the only way to pass the info to RetroArch, since it doesn't support Stella commandline arguments.

So I'm marking this one as closed.

**Comment by thrust26** [comment]
@sa666666 
Stella has autodetection if people use the known I2C.inc files.

**Comment by sa666666** [comment]
Sure, I forgot about that.  The main thing I wanted to stress is that there's no way to pass the info in on the commandline, since the Retroarch port ignores it.
